pkg/localstore/reserve.go

```diff
@@ -0,0 +1,95 @@
+// Copyright 2021 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package localstore
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/ethersphere/bee/pkg/shed"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/syndtr/goleveldb/leveldb"
+)
+
+// UnreserveBatch atomically unpins chunks of a batch in proximity order upto and including po.
+// Unpinning will result in all chunks with pincounter 0 to be put in the gc index
+// so if a chunk was only pinned by the reserve, unreserving it  will make it gc-able
+func (db *DB) UnreserveBatch(id []byte, radius uint8) error {
+	db.batchMu.Lock()
+	defer db.batchMu.Unlock()
+	var (
+		item = shed.Item{
+			BatchID: id,
+		}
+		batch = new(leveldb.Batch)
+	)
+
+	i, err := db.postageRadiusIndex.Get(item)
+	if err != nil {
+		if !errors.Is(err, leveldb.ErrNotFound) {
+			return err
+		}
+		item.Radius = radius
+		if err := db.postageRadiusIndex.PutInBatch(batch, item); err != nil {
+			return err
+		}
+		return db.shed.WriteBatch(batch)
+	}
+	oldRadius := i.Radius
+	var gcSizeChange int64 // number to add or subtract from gcSize
+	unpin := func(item shed.Item) (stop bool, err error) {
+		c, err := db.setUnpin(batch, swarm.NewAddress(item.Address))
+		if err != nil {
+			return false, fmt.Errorf("unpin: %w", err)
+		}
+
+		// if the batch is unreserved we should remove the chunk
+		// from the pull index
+		item2, err := db.retrievalDataIndex.Get(item)
+		if err != nil {
+			return false, err
+		}
+		err = db.pullIndex.DeleteInBatch(batch, item2)
```

Comment:
this is not consistent. The pullindex must include chunk batches otherwise nothing will be in the pullindex from any bin outside the neighbourhood radius